pulsar/testing/CppTester.hpp

```diff
@@ -0,0 +1,325 @@
+     *
+     * If the function throws an exception, it will be marked as a failure.
+     * If you are testing for proper exception behavior, use `test_call` instead.
+     *
+     * \exsafe_par If an exception is thrown anywhere, the
+     *             tester object remains in its original state. However,
+     *             the safety of this function with respect to the arguments
+     *             depends on the safety of those arguments.
+     *
+     * \param[in] desc description of the test we are running
+     * \param[in] should_pass True if we expect the test to match expected
+     * \param[in] expected the expected return of func
+     * \param[in] func A callable object to test
+     * \param[in] args The arguments to pass to \p func
+     */
+    template<typename R, typename Func, typename...Targs>
+    void test_return(const std::string & desc,
+                     bool should_pass,
+                     const R & expected,
+                     Func func,
+                     Targs&&...args)
+    {
+        bool success;
+
+        try {
+            R r = func(std::forward<Targs>(args)...);
+            success = (expected == r) && should_pass;
+        }
+        catch(std::exception & ex)
+        {
+            print_global_debug("Caught exception in test\n");
+            print_global_debug(ex.what());
+            success = false;
+        }
+        catch(...)
+        {
+            print_global_debug("Caught unknown exception in test\n");
+            success = false;
+        }
+
+        test(desc, success);
+    }
+
+
+    /*! \brief Tests the return value of a class member function call
+     *
+     * This is similar to test_return, however can be used to test
+     * a member function of a class instead.
+     *
+     * \exsafe_par If an exception is thrown anywhere, the
+     *             tester object remains in its original state. However,
+     *             the safety of this function with respect to the arguments
+     *             depends on the safety of those arguments.
+     *
+     * \note This version is an overload for non-const member functions
+     *
+     * \note If you are going to be testing the same member function
+     *       multiple times, consider wrapping the function via
+     *       std::bind and calling test_return directly. It will
+     *       probably be easier.
+     *
+     * \param[in] desc description of the test we are running
+     * \param[in] should_pass True if we expect the test to match expected
+     * \param[in] expected the expected return of func
+     * \param[in] func Pointer to a class member function
+     * \param[in] objptr Pointer to a class object (with which to call \p func)
+     * \param[in] args The arguments to pass to \p func
+     */
+    template<typename R1, typename R2, typename Class,
+             typename... Targs1, typename... Targs2>
+    void test_member_return(const std::string & desc,
```

Comment:
Yes and no.  I thought it was clever, but I also realize that it's ugly by most opinions.